Adding support for ROIs

juliomateoslangerak opened the pull request:
I reformated some code I had to add support for posting ROIs:
- Created a few dataclasses for the shapes
- Created a method to post a list of shapes as a ROI
- I use pydantic as a library to validate types. I like it a lot but it adds a dependency to the ezomero. It is though one of the well supported libraries of python and behind FastAPI.
- As it is, pydantic is going to cast inputs into the correct type, which could lead to confussion in some cases. If you pass a z position to a point of 3.6, it will be casted to the int 3 that OMERO is needing. We might change that behaviour easily.
- The shape dataclasses are frozen, to avoid (as far as we can) modification of the shape after saving it into OMERO. I guess we could keep track of all of this, but, is it a common need? I don't think so.

mellertd commented:
Wow this is a big contribution, thanks! I'll take a look as soon as I can and post some thoughts. Since we are so early in the project we still need to make a lot of decisions about design philosophy and such (that we frankly haven't thought too deeply about, yet), so my apologies if we have a bunch of questions/requests. But at a first glance this all looks really great.

perlman commented:
I like this and am a fan of pydantic. I had not thought of using it in this usage.

The addition of all of the ROI types really pollutes the name space. Maybe all of this should be thrown into `ezomero.roi` or similar?

juliomateoslangerak commented:
Hi,
I agree that it should be somewhere else. 
Also, more on the philosophy side of things, the omero shapewrapper is kept inside each dataclass as a _omero_shape attribute. While this works nicely, I'm not sure this is a good idea. Alternativelly, the omero_shape should be created upon posting the ROI to OMERO and then forgotten, or returned as a list of shape ids.

mellertd commented:
Here are my thoughts:

1) First, the `post_roi` function itself looks great and totally fits our style, although I would make one adjustment (more below)
2) I really, really like the Shapes classes here, and the use of pydantic looks really cool, I am just not sure this repo is the best place for them.

My reasoning for 2 is based on a couple of points. First, `pydantic` does add a dependency to the repo which is not a huge deal by any means, but it is a dependency and we are trying to avoid those that aren't already needed for `omero-py`.

The second issue I consider to be a somewhat bigger deal and has a lot to do with the state of ROIs in the Python universe at the moment. My impression is that `napari`, `scikit-image`, `omero`, and to some extent `pyimagej` all have different ways of dealing with shapes, and this addition would add one more way. The need to support and maintain `ezomero` shapes is something that concerns me.

Would it make sense to modify `post_roi` to take omero shapes directly, and maintain the shape classes in a separate library? I would certainly advise people to use the shapes here to create the omero shapes (via `shape._omero_shape`), as this code makes the creation of omero shapes much easier than using the `omero.model` objects directly.

mellertd commented:
After much thinking about this, I really can't think of a way to implement a `post_roi` in a clean way without our own classes so I am back to only the `pydantic` concern. I am not super averse to making it a dependency, but if we do we should use it throughout, not just for the shape ROIs.

Thoughts, @perlman @erickmartins ?

juliomateoslangerak commented:
Hi,
Pydantic adds, indeed, a dependency that is not crucial. It just adds, in this case, validation of the input arguments, which I think is something I, personally, could live without. Those shape classes could be replaced with standard library's dataclasses. Also, many properties that are now assigned to the shapes (colors and stroke width) could be arguments for the post_roi. All shapes of the same roi would then share those features, which I think is a fair compromise to simplify the shape classes.

One alternative, I think maybe the best, is to rely on omero-rois (https://github.com/ome/omero-rois) for the shapes and roi classes. For the moment omero-rois only implements some mask's functionalities, but we could try to get in there omero's shapes and rois.

perlman commented:
I like this code, would use it when it lands _somewhere_. `omero-rois` seems like it would be a good place.

Once it does, the features I'd like to add would bounding box and associated binary masks for volumetric ROIs.

juliomateoslangerak commented:
I made some changes with some possilbe alternatives to some of the comments:
https://github.com/juliomateoslangerak/ezomero/tree/Removed_pydantic
- removed pydantic
- put shape classes into a different module (I thought that looked cleaner)
- some features of the shapes are now added as a roi feature. That is, color and and stroke width. I thought it is a rare case to create shapes with different colors inside the same roi, so I made it an argument for the roi... Might be confusing for some though.

mellertd commented:
Awesome! Could you make a new PR with the new branch? In the meantime we'll take a good look at the code, but at first glance it looks great

mellertd commented:
It has been pointed out that it might be better to merge the 'removed_pydantic' branch into this one, to keep one PR. So maybe try that

joshmoore commented:
Just stopping by to give general encouragement (👏🏼 💧 🥘 ) and say as on https://github.com/ome/omero-rois/issues/9, if there's somewhere that you guys want to see things refactored (omero-rois, omero-py, etc.), feel free!

juliomateoslangerak commented:
Tests are failing cause I use dataclasses that are only introduced in python 3.7.

perlman commented:
I strongly believe we should track the same python versions supported by omero-py, which I think is 3.6 at the moment.

I feel OK about adding [dataclasses](https://pypi.org/project/dataclasses/) as a dependency for backward compatibility.  I think we would only need to add `"dataclasses ; python_version < '3.7'"` to `    install_requires` in `setup.py`.

juliomateoslangerak commented:
Posting a ROI across groups fails. I have to get through the group configuration. I believe the groups created might not have the right settings.

perlman commented on setup.py:
```diff
@@ -17,7 +17,8 @@
         'omero-py',
-        'numpy'
+        'numpy',
+        'dataclasses'
```
This should be 'dataclasses ; python_version < "3.7"', e.g., tell pip to install dataclasses only if python is less than 3.7

perlman replied:
https://setuptools.readthedocs.io/en/latest/userguide/dependency_management.html#id5

juliomateoslangerak replied:
Right. I thought that part was tu plainly enforce python < 3.7 and so match omero's.

perlman commented on setup.py:
```diff
@@ -17,7 +17,8 @@
+        'dataclasses'
     ],
-    python_requires='>=3.6'
+    python_requires='<3.7'
```
This needs to be combined with the dataclasses line (or alternatiely, done as an `extra_require`).

juliomateoslangerak replied:
Then I just leave this change as it was? I wouldn't know. In the examples they use 3.6 indeed.

juliomateoslangerak commented:
I suggest to create a number of different post functions for individual shapes (post_point, post_rectangle,...) and refactor current post_roi to something like post_multishape_roi. My rationale:
- ROIs are often single shaped and usage would be simpler for the user
- Slightly more functional style as, in those cases at least, you would not have to instantiate the shapes.

mellertd commented:
Thanks for your continued effort on this @juliomateoslangerak ! We are a little backed up with "day job" work at the moment but we are still very much interested in this PR

juliomateoslangerak commented:
It's my pleasure. I'm looking forward to use this package my self.
I'm trying to do some get_rois now

mellertd commented:
Thanks again, @juliomateoslangerak ! And soooo sorry this took so long. We will aim for a new pypi release soon. May make a couple of additions though.
